### PR TITLE
fix: Fix bug in MerkleDistributor that doesn't decrement remainingAmount for merkle windows correctly

### PR DESCRIPTION
**Pull request by nicholaspai**
PR applies bug fix originally discovered [here](https://github.com/across-protocol/contracts-v2/pull/193). Goal is to make this MerkleDistributor in this repo the canonical version and make the Across MerkleDistributor an extension

### CODE REVIEW

**Review by mrice32 — APPROVED** [review]

File: packages/core/contracts/merkle-distributor/implementation/MerkleDistributor.sol [hunk 1]
```diff
@@ -156,7 +156,6 @@ contract MerkleDistributor is Ownable {
                 merkleWindows[claims[nextI].windowIndex].rewardToken != currentRewardToken
                 // Next claim reward token is different than current one.
             ) {
-                merkleWindows[_claim.windowIndex].remainingAmount -= batchedAmount;
```
Comment (nicholaspai) [inline packages/core/contracts/merkle-distributor/implementation/MerkleDistributor.sol:159]: This is the bug. `remainingAmount` should be decremented for every claim processed, and now _verifyAndMarkClaimed` takes care of this

File: packages/core/contracts/merkle-distributor/implementation/MerkleDistributor.sol [hunk 2]
```diff
@@ -136,7 +136,7 @@ contract MerkleDistributor is Ownable {
      *         when any of individual `_claim`'s `amount` exceeds `remainingAmount` for its window.
      * @param claims array of claims to claim.
      */
-    function claimMulti(Claim[] memory claims) external {
+    function claimMulti(Claim[] memory claims) external virtual {
```
Comment (nicholaspai) [inline packages/core/contracts/merkle-distributor/implementation/MerkleDistributor.sol:139]: Making this `virtual` so I can extend the functionality in `across-smart-contracts-v2`

File: packages/core/contracts/merkle-distributor/implementation/MerkleDistributorInterface.sol [hunk 3]
```diff
@@ -0,0 +1,41 @@
+// SPDX-License-Identifier: GPL-3.0-only
+pragma solidity ^0.8.0;
+
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+
+/**
+ * @notice Concise list of functions in MerkleDistributor implementation that would be called by
+ * another contract, like the AcceleratingDistributor.
```
Comment (chrismaree) [inline packages/core/contracts/merkle-distributor/implementation/MerkleDistributorInterface.sol:8]:
```suggestion
 * a consuming external contract (such as the Across Protocol's AcceleratingDistributor).
```

File: packages/core/contracts/merkle-distributor/implementation/MerkleDistributor.sol [hunk 4]
```diff
@@ -170,10 +145,8 @@ contract MerkleDistributor is Ownable {
      *         will revert. It also reverts when `_claim`'s `amount` exceeds `remainingAmount` for the window.
      * @param _claim claim object describing amount, accountIndex, account, window index, and merkle proof.
      */
-    function claim(Claim memory _claim) external {
-        _verifyAndMarkClaimed(_claim);
-        merkleWindows[_claim.windowIndex].remainingAmount -= _claim.amount;
-        merkleWindows[_claim.windowIndex].rewardToken.safeTransfer(_claim.account, _claim.amount);
+    function claim(Claim memory _claim) external virtual override {
+        _executeClaim(_claim);
```
Comment (chrismaree) [inline packages/core/contracts/merkle-distributor/implementation/MerkleDistributor.sol:149]: why does this need to be an internal function like this? it's only uses here and nowhere else, right?

Comment (nicholaspai) [reply packages/core/contracts/merkle-distributor/implementation/MerkleDistributor.sol:149]: Good point will remove

File: packages/core/contracts/merkle-distributor/implementation/MerkleDistributor.sol [hunk 5]
```diff
@@ -17,34 +18,9 @@ import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
  *         multiple Merkle roots distributions with customized reward currencies.
  * @dev    The Merkle trees are not validated in any way, so the system assumes the contract owner behaves honestly.
  */
-contract MerkleDistributor is Ownable {
+contract MerkleDistributor is MerkleDistributorInterface, Ownable {
     using SafeERC20 for IERC20;
 
-    // A Window maps a Merkle root to a reward token address.
-    struct Window {
```
Comment (nicholaspai) [inline packages/core/contracts/merkle-distributor/implementation/MerkleDistributor.sol:24]: Structs refactored to `MerkleDistributorInterface`

File: packages/core/contracts/merkle-distributor/implementation/MerkleDistributor.sol [hunk 6]
```diff
@@ -136,7 +112,7 @@ contract MerkleDistributor is Ownable {
      *         when any of individual `_claim`'s `amount` exceeds `remainingAmount` for its window.
      * @param claims array of claims to claim.
      */
-    function claimMulti(Claim[] memory claims) external {
+    function claimMulti(Claim[] memory claims) external virtual override {
```
Comment (nicholaspai) [inline packages/core/contracts/merkle-distributor/implementation/MerkleDistributor.sol:115]: Made `claimMulti` overridable so AcrossMerkleDistributor can add custom logic

File: packages/core/contracts/merkle-distributor/implementation/MerkleDistributor.sol [hunk 7]
```diff
@@ -170,9 +145,8 @@ contract MerkleDistributor is Ownable {
      *         will revert. It also reverts when `_claim`'s `amount` exceeds `remainingAmount` for the window.
      * @param _claim claim object describing amount, accountIndex, account, window index, and merkle proof.
      */
-    function claim(Claim memory _claim) external {
+    function claim(Claim memory _claim) external virtual override {
```
Comment (nicholaspai) [inline packages/core/contracts/merkle-distributor/implementation/MerkleDistributor.sol:148]: Same idea as https://github.com/UMAprotocol/protocol/pull/4209/files#r1024264613

File: packages/core/contracts/merkle-distributor/implementation/MerkleDistributor.sol [hunk 8]
```diff
@@ -170,9 +145,8 @@ contract MerkleDistributor is Ownable {
...
         _verifyAndMarkClaimed(_claim);
-        merkleWindows[_claim.windowIndex].remainingAmount -= _claim.amount;
```
Comment (nicholaspai) [inline packages/core/contracts/merkle-distributor/implementation/MerkleDistributor.sol:175]: Moved logic to decrement `remainingAmount` to `_verifyAndMarkClaimed` so future devs don't forget to decrement when claiming a leaf

**Pull request merged**